src/common/SIRF.py

```diff
@@ -211,30 +213,55 @@ def axpby(self, a, b, y, out=None, **kwargs):
         '''
-        # splits axpby in 3 steps if a and b are not numbers as 
-        # pysirf.cSIRF_axpby requires them as numbers
-        if not ( isinstance(a , Number) and isinstance(b , Number) ):
-            if out is None:
-                out = y.multiply(b)
-            else:
-                y.multiply(b, out=out)
-            tmp = self.multiply(a)
-            out.add(tmp, out=out)
-            return out
+        return self.sapyb(a, y, b, out=out, **kwargs)
+
+    def sapyb(self, a, y, b, out=None, **kwargs):
+        '''
+        Addition for data containers. Can be in place.
+
+        Returns the sum of the container data with another container 
+        data viewed as vectors.
+        a: multiplier to self, can be a number or a DataContainer
+        b: multiplier to y, can be a number or a DataContainer 
+        y: DataContainer
+        out:   DataContainer to store the result to, can be self or y.
+        '''
 
         assert_validities(self, y)
-        alpha = numpy.asarray([a.real, a.imag], dtype = numpy.float32)
-        beta = numpy.asarray([b.real, b.imag], dtype = numpy.float32)
-        
-        if out is None:
-            z = self.same_object()
-            z.handle = pysirf.cSIRF_axpby \
-                (alpha.ctypes.data, self.handle, beta.ctypes.data, y.handle)
-        else:
+
+        if out is not None:
             assert_validities(self, out)
             z = out
-            try_calling(pysirf.cSIRF_axpbyAlt \
-                (alpha.ctypes.data, self.handle, beta.ctypes.data, y.handle, z.handle))
+        else:
+            z = self.copy()
+
+        
+        if isinstance(a, Number):
+            alpha = numpy.asarray([a.real, a.imag], dtype = numpy.float32)
+
+            if isinstance(b, Number):
+                beta = numpy.asarray([b.real, b.imag], dtype = numpy.float32)
+                pysirf.cSIRF_xapyb_ss_Alt(self.handle, alpha.ctypes.data, y.handle, beta.ctypes.data, z.handle)
+            else:
+                one = numpy.asarray([1.0, 0.0], dtype = numpy.float32)
+                tmp = y.multiply(b)
+                pysirf.cSIRF_xapyb_ss_Alt(self.handle, alpha.ctypes.data, tmp.handle, one.ctypes.data, z.handle)
+        else:
+            assert_validities(self, a)
+            if isinstance(b, Number):
+                one = numpy.asarray([1.0, 0.0], dtype = numpy.float32)
+                beta = numpy.asarray([b.real, b.imag], dtype = numpy.float32)
+                tmp = self.multiply(a)
+                pysirf.cSIRF_xapyb_ss_Alt(tmp.handle, one.ctypes.data, y.handle, beta.ctypes.data, z.handle)
+            else:
+                assert_validities(self, b)
+                try:
+                    try_calling(pysirf.cSIRF_xapyb_vv_Alt(self.handle, a.handle, y.handle, b.handle, z.handle))
+                except:
+                    tmp = self.multiply(a)
+                    y.multiply(b, out=z)
+                    z.add(tmp, out=z)
```

Comment:
`axpby` previously covered vector and scalar methods, but with python computing the vector case, so I don't think it's covering more cases. But yes, it is a busy method.

In terms of the users passing a data-container to `out` that is also one of the inputs, I guess we'd have to check on the ID to stop this, but as the operation is pixel-wise anyway I assumed it would be safe to allow. It doesn't change the implementation apart from the python code order avoids an intermediary step writing to out.

I think CIL currently wraps the C++ code with ctypes and passes the type of a and b as parameters. I guess this would be an option to simplify the python and matlab ends. but `sapyb` needs to be consistent in behaviour across SIRF and CIL, I'm not sure we'd want to split it in to `sapyb_vv` or `sapyb_ss` at the python end as there's algorithms where a and b could be either.

Reply:
using input container as `out` is ok - see e.g. `__iadd__`.

I still believe unnecessary copies can be avoided, but let us leave this for now - after all, as I mentioned, in SIRF we try to do all computation in C++, so `sapyb` will slim down eventually.

Reply:
The only unnecessary copy is when `a` and `b` are of different types, and the user has passed `out`, as it will pre-calculate the `vector * vector` side. Is this where you mean?

There is a specific case if the users passes `y` and `out` as the same object, where `a` is a vector and `b` a scalar. Then an intermediate step `x.multiply(a, out=out)` would overwrite `y` before it was used. So without checking on the datacontainer IDs it seems safer that where it has to calculate in steps it writes initially for a temporary container rather than `out`. Of course if it was implemented in the C++ code it would be pixel-wise and not an issue.